### PR TITLE
Web console support for health checks

### PR DESCRIPTION
**Pull request by spadgett**
https://trello.com/c/OWOhA8fv

Replaces https://github.com/openshift/origin/pull/7312, rebased for 1.3.

TODO:

- [x] Prompt if no health checks set
- [x] Improve command input and factor out into a separate directive
- [x] Drag and drop command args
- [x] Form validation
- [x] Update online-extensions.js for new doc link

### CODE REVIEW

**Comment by spadgett** [comment]
@jwforres I have drag and drop of command args working with this library:

https://github.com/a5hik/ng-sortable

Let me know how you feel about it. I was thinking we could use it for env vars as well.

**Comment by jwforres** [comment]
seems reasonably well supported, and yeah this would let us do the env var re-ordering.  sounds good to me

**Comment by spadgett** [comment]
@jwforres ready for review

File: assets/app/views/_pod-template.html [hunk 1]
```diff
@@ -3,7 +3,19 @@
...
+  detailed (optional)
+  addHealthCheckUrl (optional)
  -->
+
+ <!-- Prompt to add health checks if missing. -->
+<div ng-if="detailed && addHealthCheckUrl && !(podTemplate | hasHealthChecks)" class="alert alert-info">
+  <span class="pficon pficon-info" aria-hidden="true"></span>
+  <span ng-if="podTemplate.spec.containers.length === 1">This container has no health checks.</span>
+  <span ng-if="podTemplate.spec.containers.length > 1">Not all containers have health checks.</span>
+  Health checks ensure your application is running correctly.
```
Comment (jwforres) [inline assets/app/views/_pod-template.html]: health check is repeated like three times, feels like its getting redundant, maybe just:
This container has no health checks, which ensure your application is running correctly.

or does that read too weird...

File: assets/app/views/_pod-template.html [hunk 2]
```diff
@@ -3,7 +3,19 @@
...
+  <a ng-href="{{addHealthCheckUrl}}" class="nowrap">Add health check</a>
```
Comment (jwforres) [inline assets/app/views/_pod-template.html]: checks plural, its plural everywhere else

File: assets/app/views/edit/health-checks.html [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                    <a href="{{'application_health' | helpLink}}" target="_blank">Learn more <i class="fa fa-external-link" aria-hidden> </i></a>
+                  </div>
+                  <fieldset ng-disabled="disableInputs">
+                    <div ng-repeat="container in containers"
+                         ng-init="formName = container.name + '-form'">
+                      <h2 ng-if="containers.length > 1">Container {{container.name}}</h2>
+                      <h3>Readiness Probe</h3>
+                      <div class="help-block" ng-if="$first">
+                        A readiness probe determines if the container is ready to
+                        service requests. A failed readiness probe signals that a
```
Comment (jwforres) [inline assets/app/views/edit/health-checks.html]: ... is ready to handle requests

... failed readiness probe means

those two minor tweaks make it sound a *little* less technical

File: assets/app/views/edit/health-checks.html [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                        container should not receive any traffic from a proxy
```
Comment (jwforres) [inline assets/app/views/edit/health-checks.html]: ... from a proxy, even if it's running.

File: assets/app/views/edit/health-checks.html [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                  <div class="learn-more-block">
+                    <a href="{{'application_health' | helpLink}}" target="_blank">Learn more <i class="fa fa-external-link" aria-hidden> </i></a>
+                  </div>
+                  <fieldset ng-disabled="disableInputs">
+                    <div ng-repeat="container in containers"
+                         ng-init="formName = container.name + '-form'">
+                      <h2 ng-if="containers.length > 1">Container {{container.name}}</h2>
+                      <h3>Readiness Probe</h3>
+                      <div class="help-block" ng-if="$first">
+                        A readiness probe determines if the container is ready to
```
Comment (jwforres) [inline assets/app/views/edit/health-checks.html]: you use "determines" here and "checks" for liveness probes, be consistent either way, i'm leaning toward "checks"

**Comment by jwforres** [comment]
When I switch to TCP Socket, the Port dropdown is defaulting to what looks like "" but it says that it is a require field?

File: assets/app/views/directives/_edit-probe.html [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+          ng-pattern="/^\d+$/"
+          min="0"
+          select-on-focus
+          ng-attr-id="{{id}}-initial-delay"
+          class="form-control"
+          ng-attr-aria-describedby="{{id}}-delay-description">
+      <span class="input-group-addon">seconds</span>
+    </span>
+    <div class="help-block" ng-attr-id="{{id}}-delay-description">
+      How long to wait after the container starts to check health.
```
Comment (jwforres) [inline assets/app/views/directives/_edit-probe.html]: I read this wrong the first time, maybe "How long to wait after the container starts before checking its health."

**Comment by spadgett** [comment]
> When I switch to TCP Socket, the Port dropdown is defaulting to what looks like "" but it says that it is a require field?

That's a bug :(

**Comment by jwforres** [comment]
i think the font-weight 600 on the command args is too heavy, it looks better with normal, i know its the default for labels but since we are already changing the font-family...

**Comment by spadgett** [comment]
TCP socket port bug fixed.

**Comment by spadgett** [comment]
Updated the command font-weight.

**Comment by spadgett** [comment]
Pulled the learn more link into the help-block to tighten the spacing. @jwforres look better?

openshift_web_console

**Comment by jwforres** [comment]
@spadgett yeah thats better

**Comment by spadgett** [comment]
I'll make the same learn more link change to the set limits page.

File: assets/app/scripts/directives/util.js [hunk 7]
```diff
@@ -169,7 +169,7 @@ angular.module('openshiftConsole')
   }).directive('onEnter', function(){
     return function (scope, element, attrs) {
       element.bind("keydown keypress", function (event) {
-        if(event.which === 13 && !scope.form.$invalid) {
```
Comment (jwforres) [inline assets/app/scripts/directives/util.js]: assuming you double checked everything that was using this directive to make sure it appropriately validates?  don't we have forms where the submit button was disabled until certain required fields were filled in, those forms may not expect the submit/save function to be called

**Pull request merged**